hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieWriteHandle.java

```diff
@@ -273,4 +280,31 @@ protected static Option<IndexedRecord> toAvroRecord(HoodieRecord record, Schema
       return Option.empty();
     }
   }
+
+  protected class AppendLogWriteCallback implements HoodieLogFileWriteCallback {
+    // here we distinguish log files created from log files being appended. Considering following scenario:
+    // An appending task write to log file.
+    // (1) append to existing file file_instant_writetoken1.log.1
+    // (2) rollover and create file file_instant_writetoken2.log.2
+    // Then this task failed and retry by a new task.
+    // (3) append to existing file file_instant_writetoken1.log.1
+    // (4) rollover and create file file_instant_writetoken3.log.2
+    // finally file_instant_writetoken2.log.2 should not be committed to hudi, we use marker file to delete it.
+    // keep in mind that log file is not always fail-safe unless it never roll over
+
```

Comment:
Under this case, if we use direct marker type, the new task would got an exception when it call `preLogFileOpen`?

Reply:
> Under this case, if we use direct marker type, the new task would always get an exception when it call `preLogFileOpen` because we use `create` instead of `createIfNotExist()`? ![image](https://user-images.githubusercontent.com/1525333/246315021-8e06dad5-b8f0-4e23-9a4f-4271d1d7096c.png)

Thanks for raising this problem. It should have some problem in spark when speculation enabled. I can have a fix on that. But you mentioned always. Do you use hudi on flink? Could you pls kindly share the stacktrace of it?

Reply:
The reason why we need to use create is because early conflict detection is bound to create method rather than createIfNotExists. To maintain this feature, I changed to this method

Reply:
@guanziyue Thanks for response. 
The problem would happen in two cases:
1. Enable speculation
2. Task failed and  Driver not fail, the following retry tasks would always failed until the job failed finally.
     // The first task attempt write to log file.
    // (1) append to existing file file_instant_writetoken1.log.1
    // Then this task failed and Driver still alive, then a new task attempt would retry
    // (2) try to append to existing file file_instant_writetoken1.log.1, but would fail because the append marker already exists.

Reply:
Could you pls have a try of the latest master branch? Sorry for this mistake. It should have be solved.

Reply:
Cool

Reply:
sorry, can you guys help me understand why preLogOpen would throw here?

Reply:
oh, I get it. in hdfs like systems, if we are using direct style markers, if two diff writers try to append to same log file (either sequentially or conrrently), we will be calling append type marker for the same log file more than once. and direct style markers will fail since the marker file already exists.
is my understanding correct? 
did we make any fix on this end or not yet?
I mean, I underststand we have reverted this patch in latest master. but for MDT purpose, I am looking to see if we can re-add this patch (per log file marker).So, trying to understand any gaps or failures we need to handle before we can add per log file marker support.

Reply:
> oh, I get it. in hdfs like systems, if we are using direct style markers, if two diff writers try to append to same log file (either sequentially or conrrently), we will be calling append type marker for the same log file more than once. and direct style markers will fail since the marker file already exists. is my understanding correct? did we make any fix on this end or not yet? I mean, I underststand we have reverted this patch in latest master. but for MDT purpose, I am looking to see if we can re-add this patch (per log file marker).So, trying to understand any gaps or failures we need to handle before we can add per log file marker support.

Yes! you are correct. This was finally fixed by https://github.com/apache/hudi/pull/9003/files. 
Unfortunately, this PR was reverted by another failure. W/o MDT, FileSystem Based FileSystemView can actually 'see' some uncommitted files, like log files being written. And according to current FileGroup definition, an uncommitted log file is considered valid as long as it has a committed base instant time. Such an uncommitted file should be correctly addressed in reading because hudi can find that the instant time in log block  is invalid.
However, with this PR, we may delete an invalid log file when commit is going to finish while a reading job may require this file existed.
In theory, such an error should not occur with MDT. For FileSystem Based FileSystemView, I failed to have an idea to fix this with a short time.

Reply:
I see. so its an issue even w/ S3 like systems?
Let me go over the scenario that you are referring to so that we are on same page. 

MDT disabled.

Writer1: 
writer1 updates records in file group1 which already has a base file and 1 log file. 
writer1 writes log file2 and logfile3 (due to spark task retries). but ideally we just need only log file, i.e. log file2. 

Writer2: 
Concurrently tries to do a snapshot read from the same table concurrently. Before reconcile step for writer1 could execute, hudi returns all 3 log files (log file1, logfile2 and log file3) as part of FSView.

Writer1:
goes through the reconcile logic and deletes the extraneous log file. so log file3 is deleted.

Writer2: 
continues w/ actual read, where it hits file not found exception wrt log file3.

Reply:
one more clarification. in hdfs like systems, we can never delete any log file during reconcile right?
for eg, writer1 could have added log file1, but before writer1 can reach the reconcile step, writer2 could have appended more data to the same log file. So, writer1 can never delete any log file since a concurrent writer could have appended to it. 

From what we know, writer1's data block could be duplicate due to spark task retries, but writer2's log block could be a valid one. 
so, I feel that we can never delete any files during reconcile step in hdfs like systems (where appends are allowed). due to concurrent writers appending to an existing log file.

Reply:
> I see. so its an issue even w/ S3 like systems? Let me go over the scenario that you are referring to so that we are on same page.
> 
> MDT disabled.
> 
> Writer1: writer1 updates records in file group1 which already has a base file and 1 log file. writer1 writes log file2 and logfile3 (due to spark task retries). but ideally we just need only log file, i.e. log file2.
> 
> Writer2: Concurrently tries to do a snapshot read from the same table concurrently. Before reconcile step for writer1 could execute, hudi returns all 3 log files (log file1, logfile2 and log file3) as part of FSView.
> 
> Writer1: goes through the reconcile logic and deletes the extraneous log file. so log file3 is deleted.
> 
> Writer2: continues w/ actual read, where it hits file not found exception wrt log file3.

My bad. Let me try to make things more clearly. Actually we mentioned two problems here. Firstly, an issue fixed by HUDI-6401. Marker file plays two roles which are flag of file writing operations (creating or append) while the other is fence concurrent writing to the same file. One write op must firstly check marker file to see if it win the permission. If not, it should return a signal rather than throw exception. Here I used create method which is wrong because it throw exception when marker conflict happen. With HUDI-6401, it is totally solved after replacing create by createIfNotExists. Now concurrent writing works fine.

Reply:
> > I see. so its an issue even w/ S3 like systems? Let me go over the scenario that you are referring to so that we are on same page.
> > MDT disabled.
> > Writer1: writer1 updates records in file group1 which already has a base file and 1 log file. writer1 writes log file2 and logfile3 (due to spark task retries). but ideally we just need only log file, i.e. log file2.
> > Writer2: Concurrently tries to do a snapshot read from the same table concurrently. Before reconcile step for writer1 could execute, hudi returns all 3 log files (log file1, logfile2 and log file3) as part of FSView.
> > Writer1: goes through the reconcile logic and deletes the extraneous log file. so log file3 is deleted.
> > Writer2: continues w/ actual read, where it hits file not found exception wrt log file3.
> 
> My bad. Let me try to make things more clearly. Actually we mentioned two problems here. Firstly, an issue fixed by [HUDI-6401](https://issues.apache.org/jira/browse/HUDI-6401). Marker file plays two roles which are flag of file writing operations (creating or append) while the other is fence concurrent writing to the same file. One write op must firstly check marker file to see if it win the permission. If not, it should return a signal rather than throw exception. Here I used create method which is wrong because it throw exception when marker conflict happen. With [HUDI-6401](https://issues.apache.org/jira/browse/HUDI-6401), it is totally solved after replacing create by createIfNotExists. Now concurrent writing works fine.



> I see. so its an issue even w/ S3 like systems? Let me go over the scenario that you are referring to so that we are on same page.
> 
> MDT disabled.
> 
> Writer1: writer1 updates records in file group1 which already has a base file and 1 log file. writer1 writes log file2 and logfile3 (due to spark task retries). but ideally we just need only log file, i.e. log file2.
> 
> Writer2: Concurrently tries to do a snapshot read from the same table concurrently. Before reconcile step for writer1 could execute, hudi returns all 3 log files (log file1, logfile2 and log file3) as part of FSView.
> 
> Writer1: goes through the reconcile logic and deletes the extraneous log file. so log file3 is deleted.
> 
> Writer2: continues w/ actual read, where it hits file not found exception wrt log file3.

Let's go to current unsolved problem. It is totally about writing and reading isolation. Even we don't have concurrent writing, the problem exists. 
MDT disabled
t1, writing job start writing a log file to an FileGroup with a valid based file existed. FG base instant time t0.
t2, writing job create two log files. First is fileid_t0.log.1_1_0_1 generated by failed task and second is fileid_t0.log.1_1_0_2 generated by a successful retry task.
t3, reading job start, FileSystemView shows that FG has one base file and two log files. Will read all of them.
t4, writing job comes to reconcile step. fileid_t0.log.1_1_0_1 is invalid. delete it and finish commit.
t5, reading job want to read fileid_t0.log.1_1_0_1 but file not exists. Error occur.

Reply:
> one more clarification. in hdfs like systems, we can never delete any log file during reconcile right? for eg, writer1 could have added log file1, but before writer1 can reach the reconcile step, writer2 could have appended more data to the same log file. So, writer1 can never delete any log file since a concurrent writer could have appended to it.
> 
> From what we know, writer1's data block could be duplicate due to spark task retries, but writer2's log block could be a valid one. so, I feel that we can never delete any files during reconcile step in hdfs like systems (where appends are allowed). due to concurrent writers appending to an existing log file.

Currently, OCC model should fence any concurrent writing to the same file group. So we can actually delete log file because only one writer can generate log file to a file group.
What I'm not sure is that what should log file mechanism be like. According to my experience, allow uncommitted content leaving in hudi bring risk to the whole system. Other components must deliberately address a potential combination of invalid files and invalid log blocks. Any mistakes may result in a correctness issue. I tried to make data and meta(commit message and MDT) in hudi consistent as much as possible by clearing invalid file in reconcile step.

Reply:
hey @guanziyue : whats your slack Id. lets chat directly in hudi workspace to get consensus on resolution faster

Reply:
> hey @guanziyue : whats your slack Id. lets chat directly in hudi workspace to get consensus on resolution faster

Sent a DM to you on slack~~ My ID is Ziyue Guan

Reply:
@nsivabalan @guanziyue I'm interested in this issue. Please count me in for the discussion on this topic; I might be able to provide some input. My slack id is Jing Zhang.